src/allocation_engine/interval_tree.rs

```diff
@@ -174,40 +96,57 @@ impl NodeState {
 
 /// Internal tree node to implement interval tree.
 #[derive(Clone, Debug, PartialEq, PartialOrd, Eq, Ord)]
-pub struct InnerNode {
+struct InnerNode {
     /// Interval handled by this node.
-    key: Range,
+    pub(crate) key: Range,
```

Comment:
I'm a bit concerned about such a high usage of `pub(crate)` in this module in combination with
```rust
/// Allocation engine used by address allocator.
pub mod allocation_engine;
```
in the crate root's `lib.rs`.

Of course this just my opinion, but in my mind the question of visibility is hierarchical and should not leak "throughout" the whole crate. Meaning that Interval Tree is an implementation detail and `interval_tree.rs` should be designed from the perspective of a standalone utility module, where the provided functionality is just `pub`.

The question of how this public functionality is used by the upper modules and whether it's exported to the end user should be decided by the upper level modules, `allocation_engine` and the top level `lib.rs`.

So, I'd rather make `interval_tree`'s functionality normally public, and control the end user experience in the top level `lib.rs` by exposing what's strictly needed - it seems like only `allocation_engine::NodeState` is needed because it's used in the `Error::InvalidStateTransition` variant.